simul/platform/mininet.go

```diff
@@ -0,0 +1,372 @@
+func (m *MiniNet) Start(args ...string) error {
+	// setup port forwarding for viewing log server
+	m.started = true
+	// Remote tunneling : the sink port is used both for the sink and for the
+	// proxy => the proxy redirects packets to the same port the sink is
+	// listening.
+	// -n = stdout == /Dev/null, -N => no command stream, -T => no tty
+	var exCmd *exec.Cmd
+	redirection := fmt.Sprintf("*:%d:%s:%d", m.MonitorPort, m.ProxyAddress, m.MonitorPort)
+	login := fmt.Sprintf("%s@%s", m.Login, m.External)
+	cmd := []string{"-nNTf", "-o", "StrictHostKeyChecking=no", "-o", "ExitOnForwardFailure=yes", "-R",
+		redirection, login}
+	exCmd = exec.Command("ssh", cmd...)
+	if err := exCmd.Start(); err != nil {
+		log.Fatal("Failed to start the ssh port forwarding:", err)
+	}
+	if err := exCmd.Wait(); err != nil {
+		log.Fatal("ssh port forwarding exited in failure:", err)
+	}
+	go func() {
+		config := strings.Split(m.config, "\n")
+		sort.Strings(config)
+		log.Lvlf1("Starting simulation %s over mininet", strings.Join(config, " :: "))
+		err := SSHRunStdout(m.Login, m.External, "cd mininet_run; ./start.py list go")
+		if err != nil {
+			log.Lvl3(err)
+		}
+		m.sshMininet <- "finished"
+	}()
+
+	return nil
+}
+
+// Wait blocks on the channel till the main-process finishes.
+func (m *MiniNet) Wait() error {
+	wait := m.CloseWait
+	if wait == 0 {
+		wait = 600
+	}
+	if m.started {
+		log.Lvl3("Simulation is started")
+		select {
+		case msg := <-m.sshMininet:
+			if msg == "finished" {
+				log.Lvl3("Received finished-message, not killing users")
+				return nil
+			}
+			log.Lvl1("Received out-of-line message", msg)
+		case <-time.After(time.Second * time.Duration(wait)):
+			log.Lvl1("Quitting after ", wait/60,
+				" minutes of waiting")
+			m.started = false
+		}
+		m.started = false
+	}
+	return nil
+}
+
+// Returns the servers to use for mininet.
+func (m *MiniNet) parseServers() error {
+	hosts, err := ioutil.ReadFile(path.Join(m.mininetDir, "server_list"))
+	if err != nil {
+		return err
+	}
+	m.HostIPs = []string{}
+	for _, hostRaw := range strings.Split(string(hosts), "\n") {
+		h := strings.Replace(hostRaw, " ", "", -1)
+		if len(h) > 0 {
+			ips, err := net.LookupIP(h)
+			if err != nil {
+				// Hope it's an IP-address
+				m.HostIPs = append(m.HostIPs, h)
+			} else {
+				log.Lvl3("Found IP for", h, ":", ips[0])
+				m.HostIPs = append(m.HostIPs, ips[0].String())
+			}
+		}
+	}
+	log.Lvl3("Nodes are:", m.HostIPs)
+	return nil
+}
+
+// getHostList returns the hosts that will be used for the communication
+// and the list for `start.py`.
+func (m *MiniNet) getHostList(rc RunConfig) (hosts []string, list string, err error) {
+	hosts = []string{}
+	list = ""
+	physicalServers := len(m.HostIPs)
+	nets := make([]*net.IPNet, physicalServers)
+	ips := make([]net.IP, physicalServers)
+	for n := range nets {
+		ips[n], nets[n], err = net.ParseCIDR(fmt.Sprintf("10.%d.0.0/16", n+1))
```

Comment:
I put some comments at the beginning and in the file - does it make more sense now?